npm/vite-plugin-cypress-esm/client/moduleCache.js

```diff
@@ -0,0 +1,145 @@
+const __cypressModuleCache = new Map()
+
+const NO_REDEFINE_LIST = new Set(['prototype'])
+
+function createProxyModule (module) {
+  // What we build our module proxy off of depends on whether the module has a default export
+  // We need to be able to support `import DefaultValue from 'module'` => `const DefaultValue = __cypressModule(module)`
+  const base = module.default || module
+  let target
+
+  // Work around for the fact that a module with a default export needs to work the same way via object destructuring
+  // for this module remapping concept to work
+  // ```
+  // import TheDefault from 'module'
+  // `TheDefault` could be an object or a function
+  // ```
+  if (typeof base === 'function') {
+    target = function (...params) {
+      if (typeof target.default === 'function') {
+        return target.default.apply(this, params)
+      }
+
+      if (typeof module === 'function') {
+        return module.apply(this, params)
+      }
+    }
+  } else {
+    target = {}
+  }
+
+  const proxies = {}
+
+  const redefinePropertyDescriptors = (module, overrides) => {
+    Object.entries(Object.getOwnPropertyDescriptors(module)).forEach(([key, descriptor]) => {
+      if (NO_REDEFINE_LIST.has(key)) {
+        return
+      }
+
+      Object.defineProperty(target, key, {
+        ...descriptor,
+        ...overrides,
+      })
+
+      if (typeof descriptor.value === 'function') {
+        proxies[key] = function (...params) {
+          return target[key].apply(this, params)
+        }
+      }
+    })
+  }
+
+  if (module.default && typeof module.default !== 'function') {
+    redefinePropertyDescriptors(module.default, {
+      writable: true,
+      enumerable: true,
+    })
+  }
+
+  redefinePropertyDescriptors(module, {
+    configurable: true,
+    writable: true,
+  })
+
+  const moduleProxy = new Proxy(target, {
+    get (_, prop, receiver) {
+      const value = target[prop]
+
+      if (typeof value === 'function') {
+        const stack = new Error().stack
+
+        if (stack.includes('Sandbox.spy')) {
+          return value
+        }
+
+        return proxies[prop]
+      }
```

Comment:
Some comments here for posterity would probably be good

```suggestion
      if (typeof value === 'function') {
        // Check to see if this retrieval is coming from a sinon `spy` creation
        // If so, we want to supply the 'true' function rather than our proxied version
        // so the spy can call through to the real implementation
        const stack = new Error().stack

        if (stack.includes('Sandbox.spy')) {
          return value
        }

        // Otherwise, return our proxied function implementation
        return proxies[prop]
      }
```